pkg/controller/populators/clone-populator.go

```diff
@@ -0,0 +1,379 @@
+
+	reconciler.planner = &clone.Planner{
+		RootObjectType:  &corev1.PersistentVolumeClaimList{},
+		OwnershipLabel:  LabelOwnedByUID,
+		UIDField:        uidField,
+		Image:           clonerImage,
+		PullPolicy:      corev1.PullPolicy(pullPolicy),
+		InstallerLabels: installerLabels,
+		Client:          reconciler.client,
+		Recorder:        reconciler.recorder,
+		Controller:      reconciler.controller,
+	}
+
+	if err := addCommonPopulatorsWatches(mgr, clonePopulator, log, cdiv1.VolumeCloneSourceRef, &cdiv1.VolumeCloneSource{}); err != nil {
+		return nil, err
+	}
+
+	if err := reconciler.planner.AddCoreWatches(reconciler.log); err != nil {
+		return nil, err
+	}
+
+	return clonePopulator, nil
+}
+
+// Reconcile the reconcile loop for the PVC with DataSourceRef of VolumeCloneSource kind
+func (r *ClonePopulatorReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("PVC", req.NamespacedName)
+	log.V(1).Info("reconciling Clone Source PVC")
+
+	pvc := &corev1.PersistentVolumeClaim{}
+	if err := r.client.Get(ctx, req.NamespacedName, pvc); err != nil {
+		if k8serrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// replace this eventually
+	if pvc.Spec.DataSourceRef != nil &&
+		pvc.Spec.DataSourceRef.Namespace != nil &&
+		*pvc.Spec.DataSourceRef.Namespace != pvc.Namespace {
+		return reconcile.Result{}, r.updateClonePhaseError(ctx, pvc, fmt.Errorf("cross namespace datasource not supported yet"))
+	}
+
+	hasFinalizer := cc.HasFinalizer(pvc, cloneFinalizer)
+	isBound := pvc.Spec.VolumeName != ""
+	isDeleted := !pvc.DeletionTimestamp.IsZero()
+
+	log.V(1).Info("pvc state", "hasFinalizer", hasFinalizer, "isBound", isBound, "isDeleted", isDeleted)
+
+	if !isDeleted && !isBound {
+		return r.reconcilePending(ctx, log, pvc)
+	}
+
+	if (isBound || isDeleted) && hasFinalizer {
+		if isBound && !isClonePhaseSucceeded(pvc) {
+			log.V(1).Info("setting phase to Succeeded")
+			return reconcile.Result{}, r.updateClonePhaseSucceeded(ctx, pvc)
+		}
+
+		return r.reconcileDone(ctx, log, pvc)
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *ClonePopulatorReconciler) reconcilePending(ctx context.Context, log logr.Logger, pvc *corev1.PersistentVolumeClaim) (reconcile.Result, error) {
+	ready, _, err := claimReadyForPopulation(ctx, r.client, pvc)
+	if err != nil {
+		return reconcile.Result{}, r.updateClonePhaseError(ctx, pvc, err)
+	}
+
+	if !ready {
+		log.V(3).Info("claim not ready for population, exiting")
+		return reconcile.Result{}, r.updateClonePhasePending(ctx, pvc)
+	}
+
+	vcs, err := clone.GetVolumeCloneSource(ctx, r.client, pvc)
+	if err != nil {
+		return reconcile.Result{}, r.updateClonePhaseError(ctx, pvc, err)
+	}
+
+	if vcs == nil {
+		log.V(3).Info("dataSourceRef does not exist, exiting")
+		return reconcile.Result{}, r.updateClonePhasePending(ctx, pvc)
+	}
+
+	cs := getSavedCloneStrategy(pvc)
+	if cs == nil {
```

Comment:
Yeah, that function is kind of big, and I tried breaking it up but it got ugly(er) to me.  Problem is that the sub functions have to return data and control flow (exit early if no dataSource/or cannot compute strategy).  If you think this can be improved upon, let's do it in a subsequent pr

Reply:
okay, there will be something for you on this